src/bluetooth_linux.c

```diff
@@ -0,0 +1,548 @@
+{
+	if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "{oa{sa{sv}}}") <= 0)
+		return;
+
+	while (0 < sd_bus_message_enter_container(reply, SD_BUS_TYPE_DICT_ENTRY, "oa{sa{sv}}")) {
+		const char *ifacepath = NULL;
+		if (sd_bus_message_read_basic(reply, 'o', &ifacepath) <= 0)
+			return;
+
+		if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "{sa{sv}}") < 0)
+			return;
+		while (0 < sd_bus_message_enter_container(reply, SD_BUS_TYPE_DICT_ENTRY, "sa{sv}")) {
+			new_dbus_interface(data, ifacepath, reply);
+			sd_bus_message_exit_container(reply);
+		}
+		sd_bus_message_exit_container(reply);
+		sd_bus_message_exit_container(reply);
+	}
+}
+
+void *
+fido_bluetooth_open(const char *path)
+{
+	struct ble *newdev;
+	sd_bus_message *reply = NULL;
+	int ret;
+	if (!fido_is_bluetooth(path))
+		return NULL;
+
+	path += strlen(FIDO_BLUETOOTH_PREFIX);
+
+	newdev = calloc(1, sizeof(*newdev));
+	if (!newdev)
+		return NULL;
+
+	newdev->paths.dev = strdup(path);
+	if (!newdev->paths.dev)
+		goto out;
+
+	if (0 > sd_bus_default_system(&newdev->bus))
+		goto out;
+
+	if (0 > sd_bus_call_method(newdev->bus, "org.bluez",
+		path, "org.freedesktop.DBus.Properties", "GetAll", NULL, &reply,
+		"s", DBUS_DEV_IFACE))
+		goto out;
+
+	if (!ble_fido_is_useable_device(DBUS_DEV_IFACE, reply, false, NULL))
+		goto out;
+
+	sd_bus_message_unref(reply);
+	reply = NULL;
+	ret = sd_bus_call_method(newdev->bus, "org.bluez", "/", "org.freedesktop.DBus.ObjectManager", "GetManagedObjects", NULL, &reply, "");
+	if (ret <= 0)
+		goto out;
+
+	sd_bus_message_rewind(reply, 1);
+	iterate_over_all_objs(reply, collect_device_chars, newdev);
+
+	sd_bus_message_unref(reply);
+	reply = NULL;
+
+	if (newdev->paths.status &&
+	    newdev->paths.control_point &&
+	    newdev->paths.control_point_length &&
+	    newdev->paths.service_revision) {
+		uint8_t cp_len[2];
+		uint8_t revision;
+		if (0 > sd_bus_call_method(newdev->bus, "org.bluez", newdev->paths.control_point_length,
+					DBUS_CHAR_IFACE, "ReadValue", NULL, &reply, "a{sv}", 0))
+			goto out;
+
+		if (0 > sd_bus_message_read(reply, "ay", 2, cp_len, cp_len + 1))
+			goto out;
+
+		sd_bus_message_unref(reply);
+		reply = NULL;
+
+		if (0 > sd_bus_call_method(newdev->bus, "org.bluez", newdev->paths.service_revision,
+					DBUS_CHAR_IFACE, "ReadValue", NULL, &reply, "a{sv}", 0))
+			goto out;
+		if (0 > sd_bus_message_read(reply, "ay", 1, &revision))
+			goto out;
+
+		/* for simplicity, we allow now only FIDO2 */
+		if (!(revision & 0x20))
+			goto out;
+
+		if (0 > sd_bus_call_method(newdev->bus, "org.bluez", newdev->paths.service_revision,
+				DBUS_CHAR_IFACE, "WriteValue", NULL, NULL, "aya{sv}", 1, 0x20, 0))
+			goto out;
+
+		newdev->controlpoint_size = ((size_t)cp_len[0] << 8) + cp_len[1];
+		newdev->reply_buf = calloc(1, newdev->controlpoint_size);
+
+		if (0 > sd_bus_call_method(newdev->bus, "org.bluez", newdev->paths.status,
+					DBUS_CHAR_IFACE, "StartNotify", NULL, NULL, ""))
```

Comment:
Do you think using `AcquireNotify` instead would make sense?

We'd then be able to get rid of the `slot`, `reply_buf`, `reply_len` members of `struct ble` and be able to drop `read_cb` in its entirety. After setting up the file descriptor, we'd basically just need a call to  `fido_hid_unix_wait()`  followed by `read(2)` in `fido_bluetooth_read()`.

Reply:
yes, that is a really nice idea